### ISSUE TITLE
Version numbering

### ISSUE THREAD

**Issue by simonsan** [SFTtech/openage#1133]
We had a small discussion about that after the installer release. The version number didn't increase for a long time and so we were releasing the installer under a four years old release number. I think it is not easy to decide how and when to increase a version number. With a date-based versioning system we wouldn't really need to think about a version number anymore or if we should increase. 

YYYY - Full year 
0M - Zero-padded month

As an example or proposal: `<YYYY>.<0M><Minor>-<Modifier>`

Output:
19.061-beta
19.062-rc1
19.063-rc2

I was taking inspiration from https://calver.org/

So what do you think about a switch to date-based version numbering and if yes, what would be your favourite style then?

**Comment by heinezen** [comment]
> The version number didn't increase for a long time and so we were releasing the installer under a four years old release number.

Using calver would only make this problem more apparent to users (as we would be on 15.X right now). In my opinion, this versioning scheme is much more suited to projects with regular release schedules and where the date has some implication (e.g. for security, support or else). Also, the usage of numbers higher than 0 while we're still not finished could be confusing because it might be interpreted as the engine being feature complete.

I would personally prefer if the version number had at least some meaning and communicate the severity of changes better. In the end, the users are not interested in the version number anyway but when an update occurs, they do care about:

* Do my mods still work?
* Did a new major feature arrive?
* Do I have to adapt my mod to API changes? (modder)
* Did the behavior of an engine function change? (modder)
* Does a new feature break anything? (modder)

These questions are usually not addressed by calver as a change from 19.06 to 19.07 could include major breakages, while the releases from 19.08 to 20.02 might only handle bugfixes.

I'm much more in favor of a *feature-based* versioning using a `major.minor.patch-<tag>` structure and increase the version number in the following events:

* `major` increases for fundamental changes in the engine design's core parts (e.g. a game state rewrite, new renderer)
* `minor` increases if the API is changed in such away that it breaks older functionality, changes its structure or introduces a bunch of new features
* `patch` increases for bugfixes
* `<tag>` for nightly/alpha/beta/rc builds

This would be far better to communicate as any bump of `major` or `minor` indicates that something has changed which is either exciting (new features) or something to be aware of (API changes, engine behavior, potential breakage).

**Comment by zuntrax** [comment]
I see no reason not to use SemVer as we are already tying to do.

We have three problems here:
1. Nobody cared enough to just tag and publish new release in the last three years. (`0.3.1` was tagged some time ago, but not published.)
2. A new release was published without bumping the version number. This should not have happened. Valid versions would have been `0.3.2` or `0.4.0` at that time.
3. When I took notice of problem 2, I wanted to publish `0.4.0`, but I got distracted during setup of a windows build environment needed for that.

None of those problems would be solved or have been prevented by using another versioning scheme. In any case, proper documentation of how publishing a release works is necessary.

@Jogai 
It's pretty similar, but there are differences. SemVer makes a strong distinction between initial development releases (pre `1.0.0`) and anything afterwards. During initial development, there are no compatibility guarantees. However, for post `1.0.0`, the major number must be incremented for every breaking public API change, contrary to what @heinezen suggested. I strongly recommend reading up on the details at https://semver.org/

To sum it up, a new release should be tagged and published as soon as there is something worth releasing. Just increment the minor version with every release. I'm perfectly fine with having version `0.453.0` at some point. Don't worry about when to release `1.0.0` that's pretty far away and I will bring it up eventually.

**Comment by zuntrax** [comment]
I will meet with @TheJJ in person this evening and we will try to clean up the mess we caused.

**Comment by simonsan** [comment]
Thanks for the interesting information and discussion here. I changed the title to a more general phrasing, as that is more what it is about, maybe? I was not aware and also didn't find anything about that. When we made the installer I was unsure how it works with the version number so I was taking the one already there. I also see the points mentioned here and if there is a clear process documentation **how** and **when** to increase a version number, that would be also a solution to the insecurity at least I had about that.

**Comment by zuntrax** [comment]
@simonsan Both **how** and **when** will be addressed in #1134, which i plan to fix today.

**Comment by heinezen** [comment]
@zuntrax @Jogai 

It's deliberately deviating from SemVer because of the "breaking public API must increment major version" rule. This is better suited for libraries which almost always have backwards compatibility; something we cannot necessarily guarantee for the early openage API i fear, even after `1.0.0`.

The system I suggested is used by Paradox Games in Europa Universalis 4 and Stellaris and is very good at signalling when a major breakage can occur or when save games have the potential to stop functioning properly.

**Comment by zuntrax** [comment]
@heinezen 
I get your point, but I'd prefer not releasing `1.0.0` before our public API is somewhat stable.
I'd consider asset format, modding API, network protocol, savegame format and probably even CLI flags as part of that API. Either way, we probably shouldn't release `1.0.0` before we have proper integration tests that will fail when such breaking changes occur.
